src/Components/Shared/src/FileDownloader.cs

```diff
@@ -0,0 +1,39 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components;
+
+/// <summary>
+/// Allows triggering a file download on the client from .NET.
+/// </summary>
+internal sealed class FileDownloader : IFileDownloader
+{
+    private readonly IJSRuntime _jsRuntime;
+    public FileDownloader(IJSRuntime JSRuntime)
+    {
+        _jsRuntime = JSRuntime;
+    }
+
+    /// <inheritdoc />
+    public Task SaveAs(string fileName, byte[] data)
+    {
+        var fileStream = new MemoryStream(data);
+
+        return SaveAs(fileName, fileStream);
+    }
+
+    /// <inheritdoc />
+    public async Task SaveAs(string fileName, Stream data)
+    {
+        if (string.IsNullOrWhiteSpace(fileName))
+        {
+            throw new ArgumentNullException(nameof(fileName));
+        }
+
+        using var streamRef = new DotNetStreamReference(stream: data);
+
+        await _jsRuntime.InvokeVoidAsync("Blazor._internal.downloadFile", streamRef, fileName);
```

Comment:
@TanayParikh, could you comment on the efficiency of sending streams versus byte arrays, both for Server and WebAssembly? In my memory, on WebAssembly at least, sending a `byte[]` is super efficient because it's done in a single step without even copying the data in memory.

If that is the case, it would be advantageous to have two different internal APIs, e.g., `Blazor._internal.downloadFileArray` and `Blazor._internal.downloadFileStream`, and pick between them based on the supplied data (so we don't have to normalize both cases to streams). For Server I'm not sure whether we want to prefer `byte[]` or `Stream` if we're given a `byte[]`.

However if I'm misremembering and there isn't a perf advantage to using `byte[]` on WebAssembly then this looks good as-is!

Reply:
> In my memory, on WebAssembly at least, sending a `byte[]` is super efficient because it's done in a single step without even copying the data in memory.

That's correct, `byte[]` interop is achieved via an `InvokeUnmarshalled` call in WebAssembly, making it super efficient.

https://github.com/dotnet/aspnetcore/blob/bc3a2dd964c1e3c5aae87de7d2d5343996f487c9/src/Components/WebAssembly/JSInterop/src/WebAssemblyJSRuntime.cs#L76

https://github.com/dotnet/aspnetcore/blob/bc3a2dd964c1e3c5aae87de7d2d5343996f487c9/src/Components/Web.JS/src/Boot.WebAssembly.ts#L203

Streaming interop just leverages this unmarshalled interop for the transfer of the underlying byte chunks:
https://github.com/dotnet/aspnetcore/blob/bc3a2dd964c1e3c5aae87de7d2d5343996f487c9/src/Components/Shared/src/TransmitDataStreamToJS.cs#L24

For WebAssembly `byte[]`s can be handled separately so we can take advantage of this feature. The part which is a bit unclear is whether this performance benefit is worth having the additional complexity specifically for this case.

For Blazor Server, given the SignalR message size limitation (and assuming downloads are likely to be greater than 32kb), I don't think we're going to see too much of a benefit having a separate path for `byte[]` interop.